File: ptern-typescript/src/semantic/validator.ts
import {
  type Atom,
  type Capture,
  type Definition,
  type Exclusion,
  type Expression,
  type ParsedPtern,
  type RangeItem,
  type RepCount,
  type Repetition,
  type Sequence,
} from "../parser/ast";
import { type SemanticError } from "./error";

const KNOWN_ANNOTATIONS = new Set([
  "case-insensitive",
  "multiline",
  "replacements-ignore-matching",
  "substitutable",
  "substitutions-ignore-matching",
  "allow-backtracking",
]);

const KNOWN_POSITION_ASSERTIONS = new Set([
  "word-start",
  "word-end",
  "line-start",
  "line-end",
]);

export function validate(ptern: ParsedPtern): SemanticError[] {
  const isSubstitutable = ptern.annotations.some(a => a.name === "substitutable" && a.value);
  const defBodies = new Map<string, Expression>(ptern.definitions.map(d => [d.name, d.body]));

  const substAnnotErrs = validateSubstitutionAnnotations(ptern.annotations);
  const bodySubstErrs: SemanticError[] = isSubstitutable && !isSubstitutableExpr(ptern.body, defBodies)
    ? [{ kind: "notSubstitutableBody" }]
    : [];

  return [
    ...validateAnnotations(ptern.annotations),
    ...substAnnotErrs,
    ...bodySubstErrs,
    ...validateDefinitions(ptern.definitions),
    ...validateExpression(ptern.body, false, isSubstitutable, defBodies),
  ];
}

// ---------------------------------------------------------------------------
// Annotations
// ---------------------------------------------------------------------------

function validateAnnotations(anns: ParsedPtern["annotations"]): SemanticError[] {
  const nameErrs: SemanticError[] = anns.flatMap(ann =>
    KNOWN_ANNOTATIONS.has(ann.name) ? [] : [{ kind: "unknownAnnotation", name: ann.name } as SemanticError],
  );
  return [...nameErrs, ...findDuplicateNames(anns.map(a => a.name), "duplicateAnnotation")];
}

function validateSubstitutionAnnotations(anns: ParsedPtern["annotations"]): SemanticError[] {
  const isSubstitutable = anns.some(a => a.name === "substitutable" && a.value);
  const ignoreMatchingSet = anns.some(a => a.name === "substitutions-ignore-matching" && a.value);
  return ignoreMatchingSet && !isSubstitutable
    ? [{ kind: "substitutionsIgnoreMatchingWithoutSubstitutable" }]
    : [];
}

// ---------------------------------------------------------------------------
// Definitions
// ---------------------------------------------------------------------------

function validateDefinitions(defs: Definition[]): SemanticError[] {
  return defs.flatMap(def => validateExpression(def.body, false, false, new Map()));
}

// ---------------------------------------------------------------------------
// Expression tree walk
// ---------------------------------------------------------------------------

function validateExpression(
  expr: Expression,
  insideRep: boolean,
  isSubst: boolean,
  defBodies: Map<string, Expression>,
): SemanticError[] {
  return expr.alternatives.flatMap(seq => validateSequence(seq, insideRep, isSubst, defBodies));
}

function validateSequence(
  seq: Sequence,
  insideRep: boolean,
  isSubst: boolean,
  defBodies: Map<string, Expression>,
): SemanticError[] {
  return seq.items.flatMap(cap => validateCapture(cap, insideRep, isSubst, defBodies));
}

function validateCapture(
  cap: Capture,
  insideRep: boolean,
  isSubst: boolean,
  defBodies: Map<string, Expression>,
): SemanticError[] {
  const covered = isSubst && cap.name !== null;
  return validateRepetition(cap.inner, insideRep, isSubst, covered, defBodies);
}

function validateRepetition(
  rep: Repetition,
  insideRep: boolean,
  isSubst: boolean,
  coveredByCapture: boolean,
  defBodies: Map<string, Expression>,
): SemanticError[] {
  const countErrs: SemanticError[] = [];
  if (rep.count !== null) {
    countErrs.push(...validateRepCount(rep.count));
    const excl = rep.inner;
    if (excl.excluded === null && excl.base.kind === "singleAtom" && excl.base.atom.kind === "positionAssertion") {
      countErrs.push({ kind: "positionAssertionInRepetition", name: excl.base.atom.name });
    }
    if (isSubst && !coveredByCapture) {
      const max = rep.count.max;
      if (max.kind === "exact" || max.kind === "unbounded") {
        if (!hasNamedCaptureInExclusion(rep.inner)) {
          countErrs.push({ kind: "boundedRepetitionNeedsCapture" });
        }
      }
    }
  }
  const subInside = rep.count !== null ? true : insideRep;
  return [...countErrs, ...validateExclusion(rep.inner, subInside, isSubst, defBodies)];
}

function validateRepCount(rc: RepCount): SemanticError[] {
  const errs: SemanticError[] = [];
  if (rc.max.kind === "none" && rc.lazy) errs.push({ kind: "fewestOnExactRepetition" });
  if (rc.max.kind === "exact" && rc.min > rc.max.value) {
    errs.push({ kind: "invertedRepetitionBounds", min: rc.min, max: rc.max.value });
  }
  return errs;
}

function validateExclusion(
  excl: Exclusion,
  insideRep: boolean,
  isSubst: boolean,
  defBodies: Map<string, Expression>,
): SemanticError[] {
  const baseErrs = validateRangeItem(excl.base, insideRep, isSubst, defBodies);
  if (excl.excluded === null) return baseErrs;
  const exclErrs = validateRangeItem(excl.excluded, insideRep, isSubst, defBodies);
  const setErrs: SemanticError[] = isCharSet(excl.base, defBodies) && isCharSet(excl.excluded, defBodies)
    ? (rangeItemsEqual(excl.base, excl.excluded) ? [{ kind: "emptyCharacterSet" }] : [])
    : [{ kind: "invalidExclusionOperand" }];
  return [...baseErrs, ...exclErrs, ...setErrs];
}

function validateRangeItem(
  item: RangeItem,
  insideRep: boolean,
  isSubst: boolean,
  defBodies: Map<string, Expression>,
): SemanticError[] {
  if (item.kind === "singleAtom") return validateAtom(item.atom, insideRep, isSubst, defBodies);
  return validateCharRange(item.from, item.to);
}

function validateCharRange(from: Atom, to: Atom): SemanticError[] {
  const checkEndpoint = (atom: Atom): SemanticError[] => {
    if (atom.kind === "literal") {
      const errs: SemanticError[] = decodedLength(atom.content) !== 1
        ? [{ kind: "invalidRangeEndpoint", content: atom.content }]
        : [];
      return [...errs, ...validateLiteralEscapes(atom.content)];
    }
    return [{ kind: "invalidRangeEndpoint", content: "<non-literal>" }];
  };
  const fromErrs = checkEndpoint(from);
  const toErrs = checkEndpoint(to);
  const invErrs: SemanticError[] = [];
  if (from.kind === "literal" && to.kind === "literal" && from.content.length === 1 && to.content.length === 1) {
    if (from.content.codePointAt(0)! > to.content.codePointAt(0)!) {
      invErrs.push({ kind: "invertedRange", from: from.content, to: to.content });
    }
  }
  return [...fromErrs, ...toErrs, ...invErrs];
}

function validateAtom(
  atom: Atom,
  insideRep: boolean,
  isSubst: boolean,
  defBodies: Map<string, Expression>,
): SemanticError[] {
  switch (atom.kind) {
    case "literal":
      if (atom.content === "") return [{ kind: "emptyLiteral" }];
      return validateLiteralEscapes(atom.content);
    case "charClass":
    case "interpolation":
      return [];
    case "group":
      return validateExpression(atom.inner, insideRep, isSubst, defBodies);
    case "positionAssertion":
      return KNOWN_POSITION_ASSERTIONS.has(atom.name)
        ? []
        : [{ kind: "unknownPositionAssertion", name: atom.name }];
  }
}

// ---------------------------------------------------------------------------
// Substitutability checks
// ---------------------------------------------------------------------------

function isSubstitutableExpr(expr: Expression, defBodies: Map<string, Expression>): boolean {
  return expr.alternatives.every(seq => isSubstitutableSeq(seq, defBodies));
}

function isSubstitutableSeq(seq: Sequence, defBodies: Map<string, Expression>): boolean {
  return seq.items.every(cap => isSubstitutableCap(cap, defBodies));
}

function isSubstitutableCap(cap: Capture, defBodies: Map<string, Expression>): boolean {
  return cap.name !== null || isSubstitutableRep(cap.inner, defBodies);
}

function isSubstitutableRep(rep: Repetition, defBodies: Map<string, Expression>): boolean {
  if (rep.count === null) return isSubstitutableExcl(rep.inner, defBodies);
  if (rep.count.max.kind === "none") return isSubstitutableExcl(rep.inner, defBodies);
  return hasNamedCaptureInExclusion(rep.inner);
}

function isSubstitutableExcl(excl: Exclusion, defBodies: Map<string, Expression>): boolean {
  if (excl.excluded !== null) return false;
  return isSubstitutableItem(excl.base, defBodies);
}

function isSubstitutableItem(item: RangeItem, defBodies: Map<string, Expression>): boolean {
  if (item.kind === "charRange") return false;
  return isSubstitutableAtom(item.atom, defBodies);
}

function isSubstitutableAtom(atom: Atom, defBodies: Map<string, Expression>): boolean {
  switch (atom.kind) {
    case "literal":
    case "positionAssertion":
      return true;
    case "charClass":
      return false;
    case "interpolation": {
      const body = defBodies.get(atom.name);
      return body !== undefined && isSubstitutableExpr(body, defBodies);
    }
    case "group":
      return isSubstitutableExpr(atom.inner, defBodies);
  }
}

function hasNamedCaptureInExclusion(excl: Exclusion): boolean {
  return hasNamedCaptureInItem(excl.base);
}

function hasNamedCaptureInItem(item: RangeItem): boolean {
  return item.kind === "singleAtom" && hasNamedCaptureInAtom(item.atom);
}

function hasNamedCaptureInAtom(atom: Atom): boolean {
  return atom.kind === "group" && hasNamedCaptureInExpr(atom.inner);
}

function hasNamedCaptureInExpr(expr: Expression): boolean {
  return expr.alternatives.some(seq => hasNamedCaptureInSeq(seq));
}

function hasNamedCaptureInSeq(seq: Sequence): boolean {
  return seq.items.some(cap => hasNamedCaptureInCap(cap));
}

function hasNamedCaptureInCap(cap: Capture): boolean {
  return cap.name !== null || hasNamedCaptureInRep(cap.inner);
}

function hasNamedCaptureInRep(rep: Repetition): boolean {
  return hasNamedCaptureInExclusion(rep.inner);
}

// ---------------------------------------------------------------------------
// Character set helpers
// ---------------------------------------------------------------------------

function isSimpleCharSet(item: RangeItem): boolean {
  if (item.kind === "charRange") {
    return item.from.kind === "literal" && item.to.kind === "literal";
  }
  const atom = item.atom;
  if (atom.kind === "literal") return decodedLength(atom.content) === 1;
  return atom.kind === "charClass";
}

function isCharSet(item: RangeItem, defBodies: Map<string, Expression>): boolean {
  if (item.kind === "singleAtom") {
    const atom = item.atom;
    if (atom.kind === "group") {
      const alts = atom.inner.alternatives;
      return alts.length > 0 && alts.every(alt => isCharSetGroupAlt(alt));
    }
    if (atom.kind === "interpolation") {
      const body = defBodies.get(atom.name);
      return body !== undefined && isCharSetInterpBody(body, defBodies);
    }
  }
  return isSimpleCharSet(item);
}

function isCharSetGroupAlt(seq: Sequence): boolean {
  const items = seq.items;
  if (items.length !== 1) return false;
  const cap = items[0]!;
  return (
    cap.name === null &&
    cap.inner.count === null &&
    cap.inner.inner.excluded === null &&
    isSimpleCharSet(cap.inner.inner.base)
  );
}

function isCharSetInterpBody(expr: Expression, defBodies: Map<string, Expression>): boolean {
  return expr.alternatives.length > 0 && expr.alternatives.every(alt => isCharSetInterpAlt(alt, defBodies));
}

function isCharSetInterpAlt(seq: Sequence, defBodies: Map<string, Expression>): boolean {
  const items = seq.items;
  if (items.length !== 1) return false;
  const cap = items[0]!;
  return (
    cap.name === null &&
    cap.inner.count === null &&
    cap.inner.inner.excluded === null &&
    isCharSet(cap.inner.inner.base, defBodies)
  );
}

function rangeItemsEqual(a: RangeItem, b: RangeItem): boolean {
  if (a.kind !== b.kind) return false;
  if (a.kind === "charRange" && b.kind === "charRange") {
    return atomsEqual(a.from, b.from) && atomsEqual(a.to, b.to);
  }
  if (a.kind === "singleAtom" && b.kind === "singleAtom") return atomsEqual(a.atom, b.atom);
  return false;
}

function atomsEqual(a: Atom, b: Atom): boolean {
  if (a.kind !== b.kind) return false;
  if (a.kind === "literal" && b.kind === "literal") return a.content === b.content;
  if (a.kind === "charClass" && b.kind === "charClass") return a.name === b.name;
  return false;
}

// ---------------------------------------------------------------------------
// Escape sequence validation
// ---------------------------------------------------------------------------

function validateLiteralEscapes(content: string): SemanticError[] {
  const errs: SemanticError[] = [];
  let s = content;
  while (s.length > 0) {
    if (s[0] === "\\") {
      s = s.slice(1);
      if (s.length === 0) { errs.push({ kind: "invalidEscapeSequence", seq: "\\" }); break; }
      const c = s[0]!;
      if ("ntrAfv'\"\\".includes(c)) {
        s = s.slice(1);
      } else if (c === "u") {
        s = s.slice(5); // skip u + 4 hex digits (already validated by lexer)
      } else {
        errs.push({ kind: "invalidEscapeSequence", seq: "\\" + c });
        s = s.slice(1);
      }
    } else {
      const cp = s.codePointAt(0)!;
      s = s.slice(String.fromCodePoint(cp).length);
    }
  }
  return errs;
}

// ---------------------------------------------------------------------------
// Decoded length (escape sequences count as 1)
// ---------------------------------------------------------------------------

function decodedLength(content: string): number {
  let count = 0;
  let s = content;
  while (s.length > 0) {
    if (s[0] === "\\") {
      s = s.slice(1);
      if (s.length === 0) { count++; break; }
      const c = s[0]!;
      if (c === "u") s = s.slice(5);
      else s = s.slice(1);
      count++;
    } else {
      const cp = s.codePointAt(0)!;
      s = s.slice(String.fromCodePoint(cp).length);
      count++;
    }
  }
  return count;
}

// ---------------------------------------------------------------------------
// Duplicate detection helper
// ---------------------------------------------------------------------------

function findDuplicateNames(
  names: string[],
  kind: "duplicateAnnotation" | "duplicateDefinition" | "duplicateCapture",
): SemanticError[] {
  const seen = new Set<string>();
  const dups = new Set<string>();
  for (const name of names) {
    if (seen.has(name)) dups.add(name);
    else seen.add(name);
  }
  return [...dups].map(name => ({ kind, name }) as SemanticError);
}

// Re-export decodedLength and isCharSet for use by other modules.
export { decodedLength, isCharSet, isSimpleCharSet, hasNamedCaptureInExclusion };

File: ptern-kotlin/src/main/kotlin/io/ptern/semantic/Validator.kt
package io.ptern.semantic

import io.ptern.parser.ast.*

private val KNOWN_ANNOTATIONS = setOf(
    "case-insensitive",
    "multiline",
    "replacements-ignore-matching",
    "substitutable",
    "substitutions-ignore-matching",
    "allow-backtracking",
)

private val KNOWN_POSITION_ASSERTIONS = setOf(
    "word-start", "word-end", "line-start", "line-end",
)

object Validator {
    fun validate(ptern: ParsedPtern): List<SemanticError> {
        val isSubstitutable = ptern.annotations.any { it.name == "substitutable" && it.value }
        val defBodies = ptern.definitions.associate { it.name to it.body }
        val substAnnotErrs = validateSubstitutionAnnotations(ptern.annotations)
        val bodySubstErrs = if (isSubstitutable && !isSubstitutableExpr(ptern.body, defBodies))
            listOf(SemanticError.NotSubstitutableBody)
        else emptyList()

        return validateAnnotations(ptern.annotations) +
            substAnnotErrs +
            bodySubstErrs +
            validateDefinitions(ptern.definitions) +
            validateExpression(ptern.body, insideRep = false, isSubst = isSubstitutable, defBodies)
    }
}

// ---------------------------------------------------------------------------
// Annotations
// ---------------------------------------------------------------------------

private fun validateAnnotations(anns: List<PternAnnotation>): List<SemanticError> {
    val nameErrs = anns.flatMap { ann ->
        if (ann.name in KNOWN_ANNOTATIONS) emptyList()
        else listOf(SemanticError.UnknownAnnotation(ann.name))
    }
    return nameErrs + findDuplicateNames(anns.map { it.name }) { SemanticError.DuplicateAnnotation(it) }
}

private fun validateSubstitutionAnnotations(anns: List<PternAnnotation>): List<SemanticError> {
    val isSubstitutable = anns.any { it.name == "substitutable" && it.value }
    val ignoreMatchingSet = anns.any { it.name == "substitutions-ignore-matching" && it.value }
    return if (ignoreMatchingSet && !isSubstitutable)
        listOf(SemanticError.SubstitutionsIgnoreMatchingWithoutSubstitutable)
    else emptyList()
}

// ---------------------------------------------------------------------------
// Definitions
// ---------------------------------------------------------------------------

private fun validateDefinitions(defs: List<Definition>): List<SemanticError> =
    defs.flatMap { validateExpression(it.body, insideRep = false, isSubst = false, emptyMap()) }

// ---------------------------------------------------------------------------
// Expression tree walk
// ---------------------------------------------------------------------------

private fun validateExpression(
    expr: Expression,
    insideRep: Boolean,
    isSubst: Boolean,
    defBodies: Map<String, Expression>,
): List<SemanticError> = expr.alternatives.flatMap { validateSequence(it, insideRep, isSubst, defBodies) }

private fun validateSequence(
    seq: Sequence,
    insideRep: Boolean,
    isSubst: Boolean,
    defBodies: Map<String, Expression>,
): List<SemanticError> = seq.items.flatMap { validateCapture(it, insideRep, isSubst, defBodies) }

private fun validateCapture(
    cap: Capture,
    insideRep: Boolean,
    isSubst: Boolean,
    defBodies: Map<String, Expression>,
): List<SemanticError> {
    val covered = isSubst && cap.name != null
    return validateRepetition(cap.inner, insideRep, isSubst, covered, defBodies)
}

private fun validateRepetition(
    rep: Repetition,
    insideRep: Boolean,
    isSubst: Boolean,
    coveredByCapture: Boolean,
    defBodies: Map<String, Expression>,
): List<SemanticError> {
    val countErrs = mutableListOf<SemanticError>()
    if (rep.count != null) {
        countErrs += validateRepCount(rep.count)
        val excl = rep.inner
        if (excl.excluded == null && excl.base is RangeItem.SingleAtom && excl.base.atom is Atom.PositionAssertion) {
            countErrs += SemanticError.PositionAssertionInRepetition(excl.base.atom.name)
        }
        if (isSubst && !coveredByCapture) {
            val max = rep.count.max
            if ((max is RepUpper.Exact || max is RepUpper.Unbounded) && !hasNamedCaptureInExclusion(rep.inner)) {
                countErrs += SemanticError.BoundedRepetitionNeedsCapture
            }
        }
    }
    val subInside = rep.count != null || insideRep
    return countErrs + validateExclusion(rep.inner, subInside, isSubst, defBodies)
}

private fun validateRepCount(rc: RepCount): List<SemanticError> {
    val errs = mutableListOf<SemanticError>()
    if (rc.max is RepUpper.None && rc.lazy) errs += SemanticError.FewestOnExactRepetition
    if (rc.max is RepUpper.Exact && rc.min > rc.max.value) {
        errs += SemanticError.InvertedRepetitionBounds(rc.min, rc.max.value)
    }
    return errs
}

private fun validateExclusion(
    excl: Exclusion,
    insideRep: Boolean,
    isSubst: Boolean,
    defBodies: Map<String, Expression>,
): List<SemanticError> {
    val baseErrs = validateRangeItem(excl.base, insideRep, isSubst, defBodies)
    if (excl.excluded == null) return baseErrs
    val exclErrs = validateRangeItem(excl.excluded, insideRep, isSubst, defBodies)
    val setErrs = if (isCharSet(excl.base, defBodies) && isCharSet(excl.excluded, defBodies)) {
        if (rangeItemsEqual(excl.base, excl.excluded)) listOf(SemanticError.EmptyCharacterSet) else emptyList()
    } else {
        listOf(SemanticError.InvalidExclusionOperand)
    }
    return baseErrs + exclErrs + setErrs
}

private fun validateRangeItem(
    item: RangeItem,
    insideRep: Boolean,
    isSubst: Boolean,
    defBodies: Map<String, Expression>,
): List<SemanticError> = when (item) {
    is RangeItem.SingleAtom -> validateAtom(item.atom, insideRep, isSubst, defBodies)
    is RangeItem.CharRange -> validateCharRange(item.from, item.to)
}

private fun validateCharRange(from: Atom, to: Atom): List<SemanticError> {
    fun checkEndpoint(atom: Atom): List<SemanticError> = if (atom is Atom.Literal) {
        val lenErrs = if (decodedLength(atom.content) != 1)
            listOf(SemanticError.InvalidRangeEndpoint(atom.content)) else emptyList()
        lenErrs + validateLiteralEscapes(atom.content)
    } else {
        listOf(SemanticError.InvalidRangeEndpoint("<non-literal>"))
    }
    val fromErrs = checkEndpoint(from)
    val toErrs = checkEndpoint(to)
    val invErrs = if (from is Atom.Literal && to is Atom.Literal &&
        from.content.length == 1 && to.content.length == 1 &&
        from.content[0].code > to.content[0].code
    ) {
        listOf(SemanticError.InvertedRange(from.content, to.content))
    } else emptyList()
    return fromErrs + toErrs + invErrs
}

private fun validateAtom(
    atom: Atom,
    insideRep: Boolean,
    isSubst: Boolean,
    defBodies: Map<String, Expression>,
): List<SemanticError> = when (atom) {
    is Atom.Literal -> {
        if (atom.content.isEmpty()) listOf(SemanticError.EmptyLiteral)
        else validateLiteralEscapes(atom.content)
    }
    is Atom.CharClass, is Atom.Interpolation -> emptyList()
    is Atom.Group -> validateExpression(atom.inner, insideRep, isSubst, defBodies)
    is Atom.PositionAssertion ->
        if (atom.name in KNOWN_POSITION_ASSERTIONS) emptyList()
        else listOf(SemanticError.UnknownPositionAssertion(atom.name))
}

// ---------------------------------------------------------------------------
// Substitutability checks
// ---------------------------------------------------------------------------

private fun isSubstitutableExpr(expr: Expression, defBodies: Map<String, Expression>): Boolean =
    expr.alternatives.all { isSubstitutableSeq(it, defBodies) }

private fun isSubstitutableSeq(seq: Sequence, defBodies: Map<String, Expression>): Boolean =
    seq.items.all { isSubstitutableCap(it, defBodies) }

private fun isSubstitutableCap(cap: Capture, defBodies: Map<String, Expression>): Boolean =
    cap.name != null || isSubstitutableRep(cap.inner, defBodies)

private fun isSubstitutableRep(rep: Repetition, defBodies: Map<String, Expression>): Boolean {
    if (rep.count == null) return isSubstitutableExcl(rep.inner, defBodies)
    if (rep.count.max is RepUpper.None) return isSubstitutableExcl(rep.inner, defBodies)
    return hasNamedCaptureInExclusion(rep.inner)
}

private fun isSubstitutableExcl(excl: Exclusion, defBodies: Map<String, Expression>): Boolean {
    if (excl.excluded != null) return false
    return isSubstitutableItem(excl.base, defBodies)
}

private fun isSubstitutableItem(item: RangeItem, defBodies: Map<String, Expression>): Boolean =
    item is RangeItem.SingleAtom && isSubstitutableAtom(item.atom, defBodies)

private fun isSubstitutableAtom(atom: Atom, defBodies: Map<String, Expression>): Boolean = when (atom) {
    is Atom.Literal, is Atom.PositionAssertion -> true
    is Atom.CharClass -> false
    is Atom.Interpolation -> defBodies[atom.name]?.let { isSubstitutableExpr(it, defBodies) } ?: false
    is Atom.Group -> isSubstitutableExpr(atom.inner, defBodies)
}

// ---------------------------------------------------------------------------
// Named capture helpers
// ---------------------------------------------------------------------------

fun hasNamedCaptureInExclusion(excl: Exclusion): Boolean = hasNamedCaptureInItem(excl.base)

private fun hasNamedCaptureInItem(item: RangeItem): Boolean =
    item is RangeItem.SingleAtom && hasNamedCaptureInAtom(item.atom)

private fun hasNamedCaptureInAtom(atom: Atom): Boolean =
    atom is Atom.Group && hasNamedCaptureInExpr(atom.inner)

private fun hasNamedCaptureInExpr(expr: Expression): Boolean =
    expr.alternatives.any { hasNamedCaptureInSeq(it) }

private fun hasNamedCaptureInSeq(seq: Sequence): Boolean =
    seq.items.any { hasNamedCaptureInCap(it) }

private fun hasNamedCaptureInCap(cap: Capture): Boolean =
    cap.name != null || hasNamedCaptureInRep(cap.inner)

private fun hasNamedCaptureInRep(rep: Repetition): Boolean = hasNamedCaptureInExclusion(rep.inner)

// ---------------------------------------------------------------------------
// Character set helpers
// ---------------------------------------------------------------------------

private fun isSimpleCharSet(item: RangeItem): Boolean = when (item) {
    is RangeItem.CharRange -> item.from is Atom.Literal && item.to is Atom.Literal
    is RangeItem.SingleAtom -> when (val atom = item.atom) {
        is Atom.Literal -> decodedLength(atom.content) == 1
        is Atom.CharClass -> true
        else -> false
    }
}

fun isCharSet(item: RangeItem, defBodies: Map<String, Expression>): Boolean {
    if (item is RangeItem.SingleAtom) {
        val atom = item.atom
        if (atom is Atom.Group) {
            val alts = atom.inner.alternatives
            return alts.isNotEmpty() && alts.all { isCharSetGroupAlt(it) }
        }
        if (atom is Atom.Interpolation) {
            val body = defBodies[atom.name]
            return body != null && isCharSetInterpBody(body, defBodies)
        }
    }
    return isSimpleCharSet(item)
}

private fun isCharSetGroupAlt(seq: Sequence): Boolean {
    if (seq.items.size != 1) return false
    val cap = seq.items[0]
    return cap.name == null && cap.inner.count == null &&
        cap.inner.inner.excluded == null && isSimpleCharSet(cap.inner.inner.base)
}

private fun isCharSetInterpBody(expr: Expression, defBodies: Map<String, Expression>): Boolean =
    expr.alternatives.isNotEmpty() && expr.alternatives.all { isCharSetInterpAlt(it, defBodies) }

private fun isCharSetInterpAlt(seq: Sequence, defBodies: Map<String, Expression>): Boolean {
    if (seq.items.size != 1) return false
    val cap = seq.items[0]
    return cap.name == null && cap.inner.count == null &&
        cap.inner.inner.excluded == null && isCharSet(cap.inner.inner.base, defBodies)
}

private fun rangeItemsEqual(a: RangeItem, b: RangeItem): Boolean = when {
    a is RangeItem.CharRange && b is RangeItem.CharRange ->
        atomsEqual(a.from, b.from) && atomsEqual(a.to, b.to)
    a is RangeItem.SingleAtom && b is RangeItem.SingleAtom ->
        atomsEqual(a.atom, b.atom)
    else -> false
}

private fun atomsEqual(a: Atom, b: Atom): Boolean = when {
    a is Atom.Literal && b is Atom.Literal -> a.content == b.content
    a is Atom.CharClass && b is Atom.CharClass -> a.name == b.name
    else -> false
}

// ---------------------------------------------------------------------------
// Escape sequence validation
// ---------------------------------------------------------------------------

private fun validateLiteralEscapes(content: String): List<SemanticError> {
    val errs = mutableListOf<SemanticError>()
    var i = 0
    while (i < content.length) {
        if (content[i] == '\\') {
            i++
            if (i >= content.length) { errs += SemanticError.InvalidEscapeSequence("\\"); break }
            val c = content[i]
            if ("ntrAfv'\"\\".contains(c)) {
                i++
            } else if (c == 'u') {
                i += 5
            } else {
                errs += SemanticError.InvalidEscapeSequence("\\" + c)
                i++
            }
        } else {
            val cp = content.codePointAt(i)
            i += Character.charCount(cp)
        }
    }
    return errs
}

// ---------------------------------------------------------------------------
// Decoded length
// ---------------------------------------------------------------------------

fun decodedLength(content: String): Int {
    var count = 0
    var i = 0
    while (i < content.length) {
        if (content[i] == '\\') {
            i++
            if (i >= content.length) { count++; break }
            val c = content[i]
            if (c == 'u') i += 5 else i++
            count++
        } else {
            val cp = content.codePointAt(i)
            i += Character.charCount(cp)
            count++
        }
    }
    return count
}

// ---------------------------------------------------------------------------
// Duplicate detection helper
// ---------------------------------------------------------------------------

private fun findDuplicateNames(names: List<String>, make: (String) -> SemanticError): List<SemanticError> {
    val seen = mutableSetOf<String>()
    val dups = mutableSetOf<String>()
    for (name in names) {
        if (!seen.add(name)) dups.add(name)
    }
    return dups.map(make)
}
